### PR TITLE
hotfix/JM-7150 Juror search not available

### PR DESCRIPTION
**Pull request by akikrahman1**
### JIRA link (if applicable) ###

https://centralgovernmentcgi.atlassian.net/browse/JM-7150

### Change description ###

Juror search not available - added API to allow searching for Juror records based on filters

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[ x] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/juror/api/moj/repository/IJurorRepository.java [hunk 1]
```diff
@@ -12,28 +14,12 @@
...
                                                                                List<Integer> status);
 
-    List<Juror> findByJurorNumberInAndIsActiveAndPoolNumberAndCourt(List<String> jurorNumbers, boolean isActive,
-                                                                    String poolNumber, CourtLocation court);
-
-    Juror findByOwnerAndJurorNumberAndPoolNumber(String owner, String jurorNumber, String poolNumber);
-
-    List<Juror> findByPoolNumberAndOwnerAndIsActive(String poolNumber, String owner, boolean isActive);
-
+    JPAQuery<Tuple> fetchFilteredJurorRecords(JurorRecordFilterRequestQuery query,  List<String> courts);
```
Comment (Ben-Edwards-cgi) [inline src/main/java/uk/gov/hmcts/juror/api/moj/repository/IJurorRepository.java:24]: Don't need to pass courts here can grab from SecurityUtil.getCourts()

File: src/main/java/uk/gov/hmcts/juror/api/moj/service/JurorRecordServiceImpl.java [hunk 2]
```diff
@@ -1263,6 +1269,30 @@ public void markResponded(String jurorNumber) {
...
+                FilterJurorRecord.FilterJurorRecordBuilder builder = FilterJurorRecord.builder()
+                    .jurorNumber(tuple.get(QJuror.juror.jurorNumber))
+                    .jurorName(tuple.get(QJuror.juror.firstName.concat(" ").concat(QJuror.juror.lastName)))
+                    .postcode(tuple.get(QJuror.juror.postcode))
+                    .poolNumber(tuple.get(QJurorPool.jurorPool.pool.poolNumber))
+                    .courtName(tuple.get(QJurorPool.jurorPool.pool.courtLocation.name))
+                    .status(tuple.get(QJurorPool.jurorPool.status.statusDesc));
+                return builder.build();
+            },
+            500L
```
Comment (Ben-Edwards-cgi) [inline src/main/java/uk/gov/hmcts/juror/api/moj/service/JurorRecordServiceImpl.java:1292]: I think we have a constant for this limit some where. Maybe worth checking

File: src/integration-test/java/uk/gov/hmcts/juror/api/moj/controller/JurorRecordControllerITest.java [hunk 3]
```diff
@@ -5002,6 +5012,130 @@ void confirmJurorIdentityBureauNoAccess() throws Exception {
 
     }
 
+    @Nested
+    @DisplayName("Search for Juror records")
+    @Sql({"/db/mod/truncate.sql", "/db/JurorRecordController_searchForJurorRecords.sql"})
+    class SearchForJurorRecords {
```
[the remark this answers is not in the document]
Comment (akikrahman1) [reply src/integration-test/java/uk/gov/hmcts/juror/api/moj/controller/JurorRecordControllerITest.java:5018]: Added some more tests

File: src/main/java/uk/gov/hmcts/juror/api/moj/repository/JurorRepositoryImpl.java [hunk 4]
```diff
@@ -119,43 +59,40 @@ public List<Juror> findByJurorNumberInAndIsActiveAndPoolNumberAndCourtAndStatusI
...
-            .join(POOL_REQUEST).on(POOL_REQUEST.eq(JUROR_POOL.pool))
-            .where(POOL_REQUEST.poolNumber.eq(poolNumber))
-            .fetch();
+        JPAQuery<?> partialQuery = queryFactory.from(JUROR)
+            .join(JUROR_POOL).on(JUROR_POOL.juror.eq(JUROR));
+
+        if (!courts.contains("400")) {
+            // If the user is not a Bureau user, filter by the courts they have access to
+            partialQuery.where(JUROR_POOL.pool.courtLocation.locCode.in(courts));
+        }
```
[the remark this answers is not in the document]
Comment (akikrahman1) [reply src/main/java/uk/gov/hmcts/juror/api/moj/repository/JurorRepositoryImpl.java]: updated

File: src/main/java/uk/gov/hmcts/juror/api/moj/repository/JurorRepositoryImpl.java [hunk 5]
```diff
@@ -119,43 +59,40 @@ public List<Juror> findByJurorNumberInAndIsActiveAndPoolNumberAndCourtAndStatusI
...
+
+        if (null != query.getJurorNumber()) {
+            partialQuery.where(JUROR.jurorNumber.like(query.getJurorNumber() + "%"));
```
[the remark this answers is not in the document]
Comment (akikrahman1) [reply src/main/java/uk/gov/hmcts/juror/api/moj/repository/JurorRepositoryImpl.java]: updated

File: src/main/java/uk/gov/hmcts/juror/api/moj/repository/JurorRepositoryImpl.java [hunk 6]
```diff
@@ -119,43 +59,40 @@ public List<Juror> findByJurorNumberInAndIsActiveAndPoolNumberAndCourtAndStatusI
...
+        }
+
+        if (null != query.getJurorName()) {
+            partialQuery.where(JUROR.firstName.concat(" ").concat(JUROR.lastName).like(query.getJurorName() + "%"));
```
[the remark this answers is not in the document]
Comment (akikrahman1) [reply src/main/java/uk/gov/hmcts/juror/api/moj/repository/JurorRepositoryImpl.java]: updated

File: src/main/java/uk/gov/hmcts/juror/api/moj/repository/JurorRepositoryImpl.java [hunk 7]
```diff
@@ -119,43 +59,40 @@ public List<Juror> findByJurorNumberInAndIsActiveAndPoolNumberAndCourtAndStatusI
...
+        }
+
+        if (null != query.getPostcode()) {
+            partialQuery.where(JUROR.postcode.like(query.getPostcode() + "%"));
```
[the remark this answers is not in the document]
Comment (akikrahman1) [reply src/main/java/uk/gov/hmcts/juror/api/moj/repository/JurorRepositoryImpl.java]: updated

File: src/main/java/uk/gov/hmcts/juror/api/moj/repository/JurorRepositoryImpl.java [hunk 8]
```diff
@@ -119,43 +59,40 @@ public List<Juror> findByJurorNumberInAndIsActiveAndPoolNumberAndCourtAndStatusI
...
+        }
+
+        if (null != query.getPoolNumber()) {
+            partialQuery.where(JUROR_POOL.pool.poolNumber.like(query.getPoolNumber() + "%"));
```
[the remark this answers is not in the document]
Comment (akikrahman1) [reply src/main/java/uk/gov/hmcts/juror/api/moj/repository/JurorRepositoryImpl.java]: updated

File: src/integration-test/java/uk/gov/hmcts/juror/api/moj/controller/JurorRecordControllerITest.java [hunk 9]
```diff
@@ -5002,6 +5012,130 @@ void confirmJurorIdentityBureauNoAccess() throws Exception {
...
+                .jurorNumber("641600091")
+                .pageNumber(1)
+                .pageLimit(10)
+                .sortMethod(SortMethod.ASC)
+                .sortField(JurorRecordFilterRequestQuery.SortField.JUROR_NUMBER)
+                .build();
+
+            ResponseEntity<PaginatedList<FilterJurorRecord>> response =
+                restTemplate.exchange(new RequestEntity<>(request, httpHeaders, POST,
+                    URI.create("/api/v1/moj/juror-record/search")), new ParameterizedTypeReference<>() {});
```
[the remark this answers is not in the document]
Comment (akikrahman1) [reply src/integration-test/java/uk/gov/hmcts/juror/api/moj/controller/JurorRecordControllerITest.java]: updated

File: src/integration-test/resources/db/JurorRecordController_searchForJurorRecords.sql [hunk 10]
```diff
@@ -0,0 +1,24 @@
+DELETE FROM juror_mod.contact_log;
+DELETE FROM juror_mod.juror_history;
+DELETE FROM juror_mod.juror_audit;
+DELETE FROM juror_mod.juror_pool;
+DELETE FROM juror_mod.juror;
+DELETE FROM juror_mod.pool;
```
[the remark this answers is not in the document]
Comment (akikrahman1) [reply src/integration-test/resources/db/JurorRecordController_searchForJurorRecords.sql]: updated

File: src/main/java/uk/gov/hmcts/juror/api/moj/controller/JurorRecordController.java [hunk 11]
```diff
@@ -478,4 +481,20 @@ public void markResponded(@Valid @JurorNumber @P("juror_number") @PathVariable("
...
+    @PostMapping("/search")
+    @Operation(summary = "GET With Body", description = "Retrieve a list of all jurors, filtered by the "
+        + "provided search criteria")
+    public ResponseEntity<PaginatedList<FilterJurorRecord>> searchForJurorRecord(
+        @RequestBody @Valid JurorRecordFilterRequestQuery query) {
+
+        PaginatedList<FilterJurorRecord> poolRequests =
+            jurorRecordService.searchForJurorRecords(query);
+
+        if (poolRequests == null) {
```
[the remark this answers is not in the document]
Comment (akikrahman1) [reply src/main/java/uk/gov/hmcts/juror/api/moj/controller/JurorRecordController.java]: updated

File: src/main/java/uk/gov/hmcts/juror/api/moj/service/JurorRecordServiceImpl.java (hunk 2 above)
Comment (akikrahman1) [reply src/main/java/uk/gov/hmcts/juror/api/moj/service/JurorRecordServiceImpl.java:1290]: maybe make this an app setting?

Comment (Ben-Edwards-cgi) [reply src/main/java/uk/gov/hmcts/juror/api/moj/service/JurorRecordServiceImpl.java:1290]: From a quick search I can not see this constant anywhere so happy for this to be merged :)

**Review by Ben-Edwards-cgi — APPROVED** [review]

**Pull request merged**